### PR TITLE
Made masked losses compatible with masked nans

### PR DESCRIPTION
**Pull request by jackd**
NaNs are a great way to ensure certain values aren't used (e.g. those that are associated with masked values). This change ensures that masked values are correctly masked (set to zero, even when nan) rather than multiplied by zero (which leaves nans as nans).

This PR also (IMO) greatly simplifies the masking / weighting loss implementation. Test coverage is also improved.

### CODE REVIEW

File: keras/metrics/reduction_metrics.py [hunk 1]
```diff
@@ -13,8 +13,11 @@ def reduce_to_samplewise_values(values, sample_weight, reduce_fn, dtype):
     if sample_weight is not None:
         sample_weight = ops.cast(sample_weight, dtype=dtype)
         if mask is not None:
-            sample_weight = losses.loss.apply_mask(
-                sample_weight, mask, dtype=dtype, reduction="sum"
+            sample_weight, mask = losses.loss.squeeze_to_same_rank(
```
Comment (fchollet) [inline keras/metrics/reduction_metrics.py:16]: Why not refactor `apply_mask` instead?

File: keras/losses/loss.py [hunk 2]
```diff
@@ -126,48 +109,38 @@ def reduce_weighted_values(
...
+        return ops.sum(values)
 
+    if reduction == "sum_over_batch_size":
+        if mask is None:
+            # batch_size is the total number of elements
+            return ops.mean(values)
+        batch_size = ops.count_nonzero(mask)
+        values_sum = ops.sum(values)
+        # safe divide
+        return ops.cond(
```
Comment (fchollet) [inline keras/losses/loss.py:133]: This will add significant overhead to the computation. Is there a better way that doesn't involve a cond?

**Review by fchollet — COMMENTED** [review]
Thanks for the PR.

File: keras/losses/loss.py (hunk 2 above)
Comment (jackd) [reply keras/losses/loss.py:133]: There are plenty of ways to do essentially this, but is a `cond` on scalars that expensive? What about a `where`? They seemed the clearest, but other options would be:
- use `batch_size = maximum(batch_size, 1)` - should give identical results (batch size is an integer, and when it's 0 the numerator should be zero)
- use `batch_size = batch_size + epsilon` - closer to the current implementation, though IMO wrong.

**Comment by jackd** [comment]
@fchollet refactored to use a re-introduced `apply_mask`. Can you ellaborate on the overhead introduced by a scalar `cond`? Is there something about the specific circumstances here that make it expensive? I would have thought it would be cheaper than a `where`, which itself would be about the cost of any simple binary arithmetic operation

**Comment by fchollet** [comment]
The reason has to do with parallelization. Conditional branches are much harder to handle than a single serial op even if it has more flops.

@haifeng-jin can you advise here on how to test the performance impact of this change for a couple of standard models on GPU? I'd like to compare the `cond` implementation, the `where` implementation, and the baseline (current code).

**Comment by haifeng-jin** [comment]
@jackd here is what I did for benchmarking a PR.
You can find 6 colab notebooks here:
https://github.com/haifeng-jin/keras-benchmarking

Each of the notebooks are for one of the backends, either before the change or after the change.
You can just swap out the model part of the code and use your own model.

You will see the perf at the end of the notebook.

**Comment by jackd** [comment]
This isn't a priority for me, and I've already spent a lot longer on this than I intended to. If anyone else wants to take this up feel free, otherwise would a modified PR with just the simplified version (re-replacing masking with multiplication by zero) be accepted?

**Comment by gbaned** [comment]
Hi @fchollet Can you please assist on above [comments](https://github.com/keras-team/keras/pull/18829#issuecomment-1853368174) from @jackd. Thank you!

**Comment by dugujiujian1999** [comment]
Running on colab T4
---
- batch_norm_op_jax_after (HEAD -> 936dd1345d794e91b3883bf99dec66dc8021e7fc)
102967424/102967424 ━━━━━━━━━━━━━━━━━━━━ 6s 0us/step
101/101 ━━━━━━━━━━━━━━━━━━━━ 165s 1s/step - loss: 0.5308
training: 1111 ms/step
101/101 ━━━━━━━━━━━━━━━━━━━━ 33s 269ms/step
inferencing: 267 ms/step
- batch_norm_op_jax_before (HEAD -> 9620d23ab1d672e0611d3c9bd0f77a11a20b3204)
102967424/102967424 ━━━━━━━━━━━━━━━━━━━━ 1s 0us/step
101/101 ━━━━━━━━━━━━━━━━━━━━ 202s 1s/step - loss: 0.5062
training: 1076 ms/step
101/101 ━━━━━━━━━━━━━━━━━━━━ 41s 373ms/step
inferencing: 255 ms/step
---
- batch_norm_op_oom_torch_after (HEAD -> 936dd1345d794e91b3883bf99dec66dc8021e7fc)
101/101 ━━━━━━━━━━━━━━━━━━━━ 108s 1s/step - loss: 0.6557
414.0576171875
- batch_norm_op_oom_torch_before (HEAD -> 9620d23ab1d672e0611d3c9bd0f77a11a20b3204)
101/101 ━━━━━━━━━━━━━━━━━━━━ 109s 1s/step - loss: 0.5071
414.0576171875
---
- batch_norm_op_torch_after (HEAD -> 936dd1345d794e91b3883bf99dec66dc8021e7fc)
101/101 ━━━━━━━━━━━━━━━━━━━━ 71s 684ms/step - loss: 0.3952
training: 683 ms/step
101/101 ━━━━━━━━━━━━━━━━━━━━ 19s 189ms/step
inferencing: 189 ms/step
- batch_norm_op_torch_before (HEAD -> 9620d23ab1d672e0611d3c9bd0f77a11a20b3204)
101/101 ━━━━━━━━━━━━━━━━━━━━ 93s 545ms/step - loss: 0.5390
training: 543 ms/step
101/101 ━━━━━━━━━━━━━━━━━━━━ 18s 132ms/step
inferencing: 132 ms/step

**Comment by jackd** [comment]
@dugujiujian1999 maybe I'm missing something, but why do run with lower times have a higher times per step? e.g. `batch_norm_op_torch` training has 71s & 684ms/step before and 93s & 545ms/step. Surely they should be proportional?

**Comment by dugujiujian1999** [comment]
@jackd i use the code here:
https://github.com/haifeng-jin/keras-benchmarking/tree/main/prs

**Comment by sachinprasadhs** [comment]
@jackd , Can you please rebase the code to follow latest code structure like `/keras/src/..`

**Comment by jackd** [comment]
CBF at this point, close if nobody else wants to take over.
